src/audio/google_rtc_audio_processing.c

```diff
@@ -371,50 +372,59 @@ static int google_rtc_audio_processing_copy(struct comp_dev *dev)
 
 	buffer_stream_invalidate(cd->aec_reference, num_aec_reference_bytes);
 
-	aec_reference_buff_frag = 0;
-	for (frame = 0; frame < num_aec_reference_frames; frame++) {
-		for (channel = 0; channel < cd->num_aec_reference_channels; ++channel) {
-			src = audio_stream_read_frag_s16(&cd->aec_reference->stream,
-							 aec_reference_buff_frag + channel);
-			cd->aec_reference_buffer[cd->num_aec_reference_channels *
-				cd->aec_reference_frame_index + channel] = *src;
+	num_samples_remaining = num_aec_reference_frames * cd->num_aec_reference_channels;
+	while (num_samples_remaining) {
+		nmax = audio_stream_samples_without_wrap_s16(&cd->aec_reference->stream, ref);
+		n  = MIN(num_samples_remaining, nmax);
+		for (i = 0; i < n; i += cd->num_aec_reference_channels) {
+			j = cd->num_aec_reference_channels * cd->aec_reference_frame_index;
+			for (channel = 0; channel < cd->num_aec_reference_channels; ++channel) {
+				cd->aec_reference_buffer[j++] = *ref;
+				ref++;
```

Comment:
Same here as `cd->num_aec_reference_channels` could be lower or equal `cd->aec_reference->stream.channels` one should make sure `ref` is incremented by `cd->aec_reference->stream.channels` at the end of the loop.

In that implementation, it is only incremented by `cd->num_aec_reference_channels`.